support negative epoch_millis timestamps

rkophs opened the pull request:
_note: Several issues have been opened & closed for this regression. Assuming this PR is accepted, I'm not sure whether it's best practice to create yet another issue, or re-open an existing issue to link in this PR as the [contribution guidelines](https://github.com/elastic/elasticsearch/blob/0355469da553991f47526fbd0b0478d3b7be9989/CONTRIBUTING.md) ask for. Will update the PR accordingly once I get feedback 🙂 ._
 
Relevant issues:
 - https://github.com/elastic/elasticsearch/issues/79135
 - https://github.com/elastic/elasticsearch/issues/72123
 - https://github.com/elastic/elasticsearch/issues/40983

#### Context

ES 7.0 introduced a regression by no longer supporting negative timestamps once the epoch parsing code was migrated to use DateTimeFormatters instead of DateFormatter implementations. This regression appears to have been introduced 
when the [DateTimeFormatter replaced DateFormatter](https://github.com/elastic/elasticsearch/pull/36914) for parsing/formatting and the issue at hand appears to live exclusively in the parsing/formatting phase. Following the discussion in [40983](https://github.com/elastic/elasticsearch/issues/40983), it appears the issue at hand is added complexity in parsing/formatting epoch's in order to accommodate Java 8's date-time API representation of epoch timestamps as a tuple of: a signed epoch-seconds & a non-negative unsigned fractional time (sub second precision). 

Indeed, `Instant` uses a epochSeconds + non-negative nanosOfSecond fractional part.

The alternative is to use an `ISO 8601` string to represent negative epochs. That is not something we can do without several thousand lines of client code setup to handle the conversion at ingestion, queries & responses and is one of the single largest blockers to a Elasticsearch major version upgrade on our backend.

#### Changes

Enhance the formatting & parsing logic to support negative epoch timestamps:
 - Parsing: in the presence of a negative fractional timestamp, subtract the epochSeconds portion by 1second and increase the negative fraction by 1second (i.e. compute the absolute value of its complementary fraction) so that the fraction is always positive (.e.g. `"-4700.900ms"` -> `-4s + -700.900ms` -> `-5s + 299.100ms` -> `-5s + 299100000ns`)
 - Formatting: reverse operation to parsing. In the presence of a negative epochSeconds & non-zero fraction, Concat the results of: adding 1second to the epochSeconds, recompute complementary fraction of the fractional part.

_Many of the relevant issues allude to an edge-case that makes this difficult to handle. If this PR does not handle that edge-case, please shed light on what that edge-case is 🙇 ._

Commit:
support negative epoch_timestamps

ES 7.0 introduced a regression by no longer supporting negative
timestamps once the epoch parsing code was migrated to use
DateTimeFormatters instead of DateFormatter implementations

This change adds back the support for negative epoch timestamps

rkophs commented:
^ The contributor agreement was signed an hour before the PR was made.

adagios commented:
I would very much like that something like this is merged. The decision to not support one of the most common date formats is in my opinion very ill advised.

The added interface code that ES now needs when receiving things from other systems, because it doesn't support one of the simplest standards is difficult to accept as a good decision.

Imagine if Java removed the parsing of timestamps...

elasticmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

rjernst commented:
It has been many years so the details are a bit hazy for me. However, IIRC, the problem was not in the representation (which I believe I tried the same offsetting you do to make nanos of second always positive) but instead in Java’s format for negative fractional seconds. I think it was something like -0.12345, the negative is lost by the time the nanos are parsed. Can you add a test for this case?

br3no commented:
@rkophs there's an issue to address the same problem in OpenSearch here: https://github.com/opensearch-project/OpenSearch/issues/1991. Would you like to submit your pull request to OpenSearch? Otherwise, would you mind if I port your fix to OpenSearch?

rjernst commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into support_negative_epoch_millis_timestamps

rjernst commented:
@elasticmachine ok to test

Commit:
support negative fractional epochs where seconds is 0.

Java DateTimeFormatter isn't able to parse "-0". Therefore,
the MILLIS portion of the parser fails for epochs such as
"-0.12345". We can overcome this by introducing an optional
TemporalField that represents whether the timestamp is
signed. It is only created during parsing in the presence
of a leading '-' char. Formatting will thus only support it
if we're doing with a INSTANT_SECONDS temporalAccessor that
happens to be negative.

Commit:
Merge branch 'support_negative_epoch_millis_timestamps' of https://github.com/rkophs/elasticsearch into support_negative_epoch_millis_timestamps

rkophs commented:
@rjernst -- thanks for the insight. I added a test case for `-0.12345`. Indeed, you are correct. The `DateTimeFormatter` doesn't know how to parse "-0" for the millis portion of the timestamp.

I've added a bit of a "hack" to get that case to work by introducing a new `TemporalField` whose optional presence denotes that the timestamp is _signed_. This information is then able to be used by the `MILLIS` TemporalField to calculate the appropriate seconds/nanos parts during parsing. On the flipside, for formatting, the new `TemporalField` is only marked as supported for a `TemporalAccessor` that has a negative INSTANT_SECONDS value.

See my latest comment for details.

Commit:
checkstyle - replace "!" with explicit "== false"

Commit:
avoid using Math.abs

rjernst commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into support_negative_epoch_millis_timestamps

rjernst commented:
@rkophs Could you please sign the [CLA](https://www.elastic.co/contributor-agreement) and I will review this more thoroughly?

rkophs commented:
@rjernst -- I've signed that CLA several times (using every one of my several different emails registered in Github). Does Elastic's CLA automation require the email to match what's tied to the commits, etc?

I'm happy to email Elastic a copy of my signed agreement(s).

rjernst reviewed (changes requested):
@rkophs This is a very clever approach, using a dummy field as a placeholder for negative values.

I left a couple minor comments, but the one thing I think must be addressed is that adding this support just for epoch_millis is going to make support for epoch timestamps inconsistent. Your approach should work for the `epoch_second` parser as well, which is in the same EpochTime class. Would you please adjust its implementation similarly?

rkophs commented on server/src/main/java/org/elasticsearch/common/time/EpochTime.java:
```diff
@@ -90,19 +124,47 @@ public TemporalAccessor resolve(
             if (fieldValues.containsKey(ChronoField.MILLI_OF_SECOND)) {
                 fieldValues.put(ChronoField.MILLI_OF_SECOND, nanos / 1_000_000);
             }
+            if (fieldValues.containsKey(ChronoField.MICRO_OF_SECOND)) {
```
It is not needed. I added it defensively since we're the same logic is happening above with `MILLI_OF_SECOND`, but I'm happy to remove it since it's not directly related to the change at hand.

rkophs commented on server/src/main/java/org/elasticsearch/common/time/EpochTime.java:
```diff
@@ -73,15 +92,30 @@ public long getFrom(TemporalAccessor temporal) {
+                long nanos = temporal.getLong(ChronoField.NANO_OF_SECOND);
+                if (nanos % 1_000_000 != 0) {
+                    // Fractional negative timestamp.
+                    // Add 1 ms towards positive infinity because the fraction leads
```
You're understanding is correct. If we get into this branch, it means two things: 1) we're dealing with a negative epoch, 2) we're dealing with a fractional millisecond. Thus, the calculated `millis` invokes a precision loss that needs to be rounded to the _more negative_ millisecond (i.e. the millis value with the higher magnitude).

I'll update the comment to reflect.

Commit:
comment on negative epoch rounding; remove unnecessary code

Commit:
support negative epochs for epoch_seconds

rjernst commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into support_negative_epoch_millis_timestamps

rjernst commented:
Thanks for the updates @rkophs. This looks good, I will wait for CI to complete and then merge.

mark-vieira commented:
@elasticsearchmachine generate changelog

mark-vieira commented:
@elasticsearchmachine generate changelog

rjernst committed:
add changelog

rjernst committed:
fix changelog format

rjernst commented:
Thanks for your work here @rkophs! This is a clever approach.
